### PR TITLE
Add json ingest from structs and lists, and change extract to use output...

### PR DESCRIPTION
**Pull request by FrontierOtr**
... json structs and lists

### CODE REVIEW

File: ivory-storage/src/main/scala/com/ambiata/ivory/storage/parse/EavtParsers.scala [hunk 1]
```diff
@@ -48,7 +48,7 @@ object EavtParsers {
   def valueFromString(meta: ConcreteDefinition, raw: String): Validation[String, Value] = meta.encoding match {
```
Comment (markhibberd) [inline ivory-storage/src/main/scala/com/ambiata/ivory/storage/parse/EavtParsers.scala]: In this file, I don't think it is calling split on delimiter, I don't think this is safe any more, it should count forward delims for entity and attribute, and backwards for time, and treat the middle as the value (due to potential | in json.

File: ivory-core/src/main/scala/com/ambiata/ivory/core/Fact.scala [hunk 2]
```diff
@@ -217,6 +229,46 @@ object Value {
...
+        ls <- sub match {
+          case se: StructEncoding    => jl.traverseU(v => parseJsonStruct(se, v))
+          case pe: PrimitiveEncoding => jl.traverseU(v => parseJsonPrimitive(pe, v))
+        }
+      } yield ListValue(ls)
+    }
+    (for {
+      json  <- raw.decodeOption[Json].toRightDisjunction(s"Value '$raw' is not a List")
+      value <- encoding match {
+        case pe: PrimitiveEncoding => parseJsonPrimitive(pe, json)
```
Comment (markhibberd) [inline ivory-core/src/main/scala/com/ambiata/ivory/core/Fact.scala]: *shakes fist at @charleso* I don't think I can ever forgive you these interfaces.

File: ivory-storage/src/main/scala/com/ambiata/ivory/storage/parse/EavtParsers.scala [hunk 3]
```diff
@@ -48,7 +48,7 @@ object EavtParsers {
...
     case _ if meta.tombstoneValue.contains(raw)  => TombstoneValue.success[String]
     case p: PrimitiveEncoding                    => Value.parsePrimitive(p, raw)
-    case s: StructEncoding                       => "Struct encoding not supported".failure
-    case _: ListEncoding                         => "List encoding not supported".failure
+    case s: StructEncoding                       => Value.parseJson(s, raw)
+    case l: ListEncoding                         => Value.parseJson(l, raw)
```
Comment (markhibberd) [inline ivory-storage/src/main/scala/com/ambiata/ivory/storage/parse/EavtParsers.scala]: Is there an arbitrary & spec to update for these as well?

**Comment by markhibberd** [comment]
Need to add something to release notes about breaking output format. I need to think a bit more about if it is ok to just change the output format like that, it will affect some of our downstream code, but I will need to talk to @blever if it will hit anyone else (I don't think so though).

File: ivory-core/src/main/scala/com/ambiata/ivory/core/Fact.scala [hunk 4]
```diff
@@ -217,6 +229,46 @@ object Value {
     case StringEncoding                          => StringValue(raw).success[String]
   }
 
+  def parseJson(encoding: Encoding, raw: String): Validation[String, Value] = {
```
Comment (charleso) [inline ivory-core/src/main/scala/com/ambiata/ivory/core/Fact.scala]: As mentioned IRL I might be tempted to extract this to it's own file/class.

File: ivory-storage/src/main/scala/com/ambiata/ivory/storage/parse/EavtParsers.scala (hunk 3 above)
Comment (charleso) [inline ivory-storage/src/main/scala/com/ambiata/ivory/storage/parse/EavtParsers.scala]: > Is there an arbitrary & spec to update for these as well?

+1

**Comment by charleso** [comment]
@markhibberd I've pushed up what I have for feedback (and builds). Just digging in to the keyed_set problem now.

**Comment by FrontierOtr** [comment]
Did you squash my commits out? 
:(

**Pull request merged**